staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filter-config.go

```diff
@@ -0,0 +1,349 @@
+				continue
+			}
+			if plState.emptyHandler == nil {
+				klog.V(3).Infof("Priority level %q became undesired", plName)
+				plState.emptyHandler = &emptyRelay{reqMgr: reqMgr}
+				newlyQuiescent = append(newlyQuiescent, plState)
+			}
+		}
+		if plState.config.Limited != nil {
+			shareSum += float64(plState.config.Limited.AssuredConcurrencyShares)
+		} else {
+			haveExemptPL = true
+		}
+		haveCatchAllPL = haveCatchAllPL || plName == rmtypesv1a1.PriorityLevelConfigurationNameCatchAll
+		newRMState.priorityLevelStates[plName] = plState
+	}
+
+	if !haveExemptPL {
+		newRMState.imaginePL(fcboot.MandatoryPriorityLevelConfigurationExempt, reqMgr.requestWaitLimit, &shareSum)
+	}
+	if !haveCatchAllPL {
+		newRMState.imaginePL(fcboot.MandatoryPriorityLevelConfigurationCatchAll, reqMgr.requestWaitLimit, &shareSum)
+	}
+	if !haveExemptFS {
+		fsSeq = append(apihelpers.FlowSchemaSequence{fcboot.NewFSAllGroups(rmtypesv1a1.FlowSchemaNameExempt, rmtypesv1a1.PriorityLevelConfigurationNameExempt, 1, "", user.SystemPrivilegedGroup)}, fsSeq...)
+	}
+	if !haveCatchAllFS {
+		fsSeq = append(fsSeq, fcboot.NewFSAllGroups(rmtypesv1a1.FlowSchemaNameCatchAll, rmtypesv1a1.PriorityLevelConfigurationNameCatchAll, rmtypesv1a1.FlowSchemaMaxMatchingPrecedence, rmtypesv1a1.FlowDistinguisherMethodByUserType, user.AllUnauthenticated, user.AllAuthenticated))
+	}
+
+	newRMState.flowSchemas = fsSeq
+	if klog.V(5) {
+		for _, fs := range fsSeq {
+			klog.Infof("Using FlowSchema %s: %#+v", fs.Name, fs.Spec)
+		}
+	}
+	for plName, plState := range newRMState.priorityLevelStates {
+		if plState.config.Limited == nil {
+			klog.V(5).Infof("Using exempt priority level %q: quiescent=%v", plName, plState.emptyHandler != nil)
+			continue
+		}
+
+		plState.qsConfig.ConcurrencyLimit = int(math.Ceil(float64(reqMgr.serverConcurrencyLimit) * float64(plState.config.Limited.AssuredConcurrencyShares) / shareSum))
+		metrics.UpdateSharedConcurrencyLimit(plName, plState.qsConfig.ConcurrencyLimit)
+
+		if plState.queues == nil {
+			klog.V(5).Infof("Introducing queues for priority level %q: config=%#+v, concurrencyLimit=%d, quiescent=%v (shares=%v, shareSum=%v)", plName, plState.config, plState.qsConfig.ConcurrencyLimit, plState.emptyHandler != nil, plState.config.Limited.AssuredConcurrencyShares, shareSum)
+			plState.queues = reqMgr.queueSetFactory.NewQueueSet(plState.qsConfig)
+		} else {
+			klog.V(5).Infof("Retaining queues for priority level %q: config=%#+v, concurrencyLimit=%d, quiescent=%v (shares=%v, shareSum=%v)", plName, plState.config, plState.qsConfig.ConcurrencyLimit, plState.emptyHandler != nil, plState.config.Limited.AssuredConcurrencyShares, shareSum)
+			plState.queues.SetConfiguration(plState.qsConfig) // TODO: make sure that retains other config if zero desired queues
+		}
+	}
+	reqMgr.curState.Store(newRMState)
+	klog.V(5).Infof("Switched to new RequestManagementState")
+	// We do the following only after updating curState to guarantee
```

Comment:
AFAICT this does not actually happen after updating curState, it happens concurrently (from the perspective of another goroutine).

Reply:
The comment is trying to explain a longer chain of events, the concern here is not simply a race between Store and Quiesce.

Reply:
I added a second attempt to explain.

Reply:
I think the proof of safety is pretty simple; it goes as follows.

1. `curState.Store(X)` causally precedes `Y.Quiesce(non-nil)`

2. `Y.Quiesce(non-nil)` causally precedes `Y.Wait(..)` returning with `tryAnother==true`.

3. `Y.Wait(..)` returning with `tryAnother==true` causally precedes a consequent call to `curState.Load()`.  Thus, this Load returns `X` or a value that was Stored after `X`.

The logic for digesting the config objects in not quite as simple as processing each one independently; there is also the logic that adds missing objects.  It certainly _is_ possible to incrementalize this code, but it would have a bit of hair.

Given that this code is in the request path I thought it was worth minimizing the synchronization overhead.  As I said, I think the solution with immutable data structures is viable.

Reply:
Your proof of safety crosses four functions in three files, with a lock in one object (the QueueSet) used to synchronize access to an object that points to it (the requestManager) through two levels of indirection (reqMgr.curState.plState.queues). I would not call that simple; it also has no static enforcement, so a relatively small future change could create subtle breakage. I don't trust myself not to break it in the future, even having read it now, and I think ideally this code should be maintainable by programmers as bad as I am :)

As for synchronization overhead in the request path... I haven't profiled apiserver, but I don't think I'm going too far out on a limb by saying that one extra RLock/RUnlock pair is not going to be noticeable there. There's much, much lower-hanging fruit on this tree, and I'm a lot more worried about safety of this feature. I'm not an owner here, though, so I won't belabor the point (much) more.

Reply:
Revised code confines the proof to one file plus imported abstractions.